src/bookshelf/src/bksw_wm_sql_object.erl

```diff
@@ -0,0 +1,336 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil; fill-column: 92 -*-
+%% ex: ts=4 sw=4 et
+%% @author Mark Anderson <mark@chef.io>
+%% @author Tim Dysinger <dysinger@opscode.com>
+%% Copyright 2012-2015 Opscode, Inc. All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(bksw_wm_sql_object).
+
+-include_lib("mixer/include/mixer.hrl").
+-mixin([{bksw_wm_base, [init/1,
+                        is_authorized/2,
+                        finish_request/2,
+                        service_available/2]}]).
+
+%% Webmachine callbacks
+-export([allowed_methods/2,
+         content_types_accepted/2,
+         content_types_provided/2,
+         delete_resource/2,
+         generate_etag/2,
+         last_modified/2,
+         resource_exists/2,
+
+         %% Override
+         validate_content_checksum/2,
+
+         %% Resource helpers
+         download/2,
+         upload/2]).
+
+
+-include_lib("webmachine/include/webmachine.hrl").
+-include("internal.hrl").
+
+%%===================================================================
+%% Public API
+%%===================================================================
+
+%% By default, if wm sees a 'content-md5' header, it will read the request body to compute
+%% the md5 and compare to the header value. A 400 will then be returned automagically by wm
+%% if the digests do not match. Since we wish to read request bodies in a streaming fashion,
+%% we need to handle our own checksum validation. Using wm's default would mean having a
+%% full copy of the request body buffered into the request process state. So we define this
+%% resource callback to blindly say the content is valid and then do the verification in the
+%% upload/2 flow.
+validate_content_checksum(Rq, Ctx) ->
+    {true, Rq, Ctx}.
+
+allowed_methods(Rq, Ctx) ->
+    {['HEAD', 'GET', 'PUT', 'DELETE'], Rq, Ctx}.
+
+content_types_provided(Rq, Ctx) ->
+    CType =
+    case wrq:get_req_header("accept", Rq) of
```

Comment:
The whitespace is off here. Perhaps we should make this method consistent with `content_types_accepted`